pkg/chain/ethereum/beacon.go

```diff
@@ -169,3 +201,270 @@ func (bc *BeaconChain) UpdateOperatorStatus() error {
 	return err
 }
+
+// SelectGroup returns the group members for the group generated by
+// the given seed. This function can return an error if the beacon chain's
+// state does not allow for group selection at the moment.
+func (bc *BeaconChain) SelectGroup(seed *big.Int) ([]chain.Address, error) {
+	groupSize := big.NewInt(int64(bc.chainConfig.GroupSize))
+	seedBytes := [32]byte{}
+	seed.FillBytes(seedBytes[:])
+
+	operatorsIDs, err := bc.sortitionPool.SelectGroup(groupSize, seedBytes)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"cannot select group in the sortition pool: [%v]",
+			err,
+		)
+	}
+
+	operatorsAddresses, err := bc.sortitionPool.GetIDOperators(operatorsIDs)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"cannot convert operators' IDs to addresses: [%v]",
+			err,
+		)
+	}
+
+	result := make([]chain.Address, len(operatorsAddresses))
+	for i := range result {
+		result[i] = chain.Address(operatorsAddresses[i].String())
+	}
+
+	return result, nil
+}
+
+func (bc *BeaconChain) OnGroupRegistered(
+	handler func(groupRegistration *event.GroupRegistration),
+) subscription.EventSubscription {
+	// TODO: Implementation.
+	panic("not implemented yet")
+}
+
+// TODO: Implement a real IsGroupRegistered function.
+func (bc *BeaconChain) IsGroupRegistered(groupPublicKey []byte) (bool, error) {
+	return false, nil
+}
+
+func (bc *BeaconChain) IsStaleGroup(groupPublicKey []byte) (bool, error) {
+	// TODO: Implementation.
+	panic("not implemented yet")
+}
+
+func (bc *BeaconChain) GetGroupMembers(
+	groupPublicKey []byte,
+) ([]chain.Address, error) {
+	// TODO: Implementation.
+	panic("not implemented yet")
+}
+
+// TODO: Implement a real OnDKGStarted event subscription. The current
+//       implementation generates a fake event every 500th block where the
+//       seed is the keccak256 of the block number.
+func (bc *BeaconChain) OnDKGStarted(
+	handler func(event *event.DKGStarted),
+) subscription.EventSubscription {
+	ctx, cancelCtx := context.WithCancel(context.Background())
+	blocksChan := bc.blockCounter.WatchBlocks(ctx)
+
+	go func() {
+		for {
+			select {
+			case block := <-blocksChan:
+				// Generate an event every 500th block.
+				if block%500 == 0 {
+					// The seed is keccak256(block).
+					blockBytes := make([]byte, 8)
+					binary.BigEndian.PutUint64(blockBytes, block)
+					seedBytes := crypto.Keccak256(blockBytes)
+					seed := new(big.Int).SetBytes(seedBytes)
+
+					handler(&event.DKGStarted{
+						Seed:        seed,
+						BlockNumber: block,
+					})
+				}
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	return subscription.NewEventSubscription(func() {
+		cancelCtx()
+	})
+}
+
+// TODO: Implement a real SubmitDKGResult action. The current implementation
+//       just creates and pipes the DKG submission event to the handlers
+//       registered in the dkgResultSubmissionHandlers map. Consider getting
+//       rid of the result promise in favor of the fire-and-forget style.
```

Comment:
Yep, we need it for the test loop because `SubmitDKGResult` have to cause an action on the `OnDKGResultSubmitted` event listener. I think it will be much easier to implement the fire-and-forget approach once we integrate the result submission with the actual `RandomBeacon` contract. This is why that TODO is here :)

Reply:
I would really love to keep the new implementation clean so that we can better navigate through the code and see what else is yet to be done.

I am not sure I follow the explanation. `SubmitDKGResult` can cause an action on `OnDKGResultSubmitted` without the promise because triggering this action is just a loop over registered handlers:

```
for _, handler := range bc.dkgResultSubmissionHandlers 
```

`submission.go` needs to be reworked anyway when we'll be integrating it with the real beacon contract. Looking at `submission.go`, the only thing the promise is doing is sending an error to a channel and I think this can be done by just returning an error from `SubmitDkgResult`. so maybe instead of adding promises to v2 chain, we could do in `beacon.go`:

```
return chainRelay.SubmitDKGResult(
  sm.index,
  result,
  signatures,
)				
```

Reply:
Yeah, my bad. Got rid of the promise: https://github.com/keep-network/keep-core/pull/3076/commits/601f8d66272e3917085e7dd3adcf45cb61a455a1